content/blog/2021-11-10-agent.md

```diff
@@ -0,0 +1,156 @@
+
+* Scrape across the network is a terrible idea as we employ a huge amount of unknowns in our monitoring pipeline. The local pull model allows Prometheus to know why exactly the metric target has problems and when. Maybe it's down, misconfigured, restarted, too slow to give us metrics (e.g. CPU saturated), not discoverable by service discovery, we don't have credentials to access or just DNS, network, or whole cluster is down. By putting our scraper outside of the network, we risk losing some of the information due to unreliable scrape and irregular scrape intervals. On top of that, we risk losing important visibility completely if the network is temporarily down. Please don't do it. It's not worth it. (:
+* Metric push directly from the application to some central location is equally bad. Especially when you monitor a larger fleet, you know nothing literally when you don't see metrics from remote applications. Is the application down? Is my receive pipeline down? Maybe the application failed to authorize? Maybe it failed to get the IP address of my remote cluster? Maybe it's too slow? Maybe the network is down? Such design is often a recipe for a failure (or at least shrug 🤷🏽‍♀️ moments).
+
+Prometheus introduced three main protocols to support the global view case, as discussed many times at various conferences. Each with its own pros and cons. Let's go briefly go through those. They are presented in orange colour on the diagram below.
+
+![Prometheus global view](/assets/blog/2021-11-10/prom-remote.png)
+
+* **Federation** was introduced as the first feature for aggregation purposes. It allows a special scrape done by Prometheus on a global level for a subset of metrics in the leaf Prometheus. Such a "federation" scrape reduces some unknowns across networks because metrics exposed by federate endpoints inject timestamps. Yet, it usually suffers from the inability to federate all metrics and not lose data during longer network partitions (minutes).
+* **Remote Read** allows to select metrics from the Prometheus database without the PromQL engine directly. You can deploy Prometheus or other solutions (e.g. Thanos) on the global level to perform PromQL queries there while fetching required metrics from multiple remote locations. This is really powerful as it allows you to store data "locally" and access it only when needed. Unfortunately, there are cons too. Without features like [Query Pushdown](https://github.com/thanos-io/thanos/issues/305) we are in extreme cases pulling GBs of compressed metric data to answer a single query. Also, if we have a network partition, we are temporarily blind. Last but not least, certain security guidelines are not allowing ingress traffic, only egress one.
+* Finally, we have **Remote Write**, which seems to be the most popular choice nowadays. Since the agent mode focuses on remote write feature use cases, let's explain it in more detail.
```

Comment:
```suggestion
* Finally, we have **Remote Write**, which seems to be the most popular choice nowadays. Since the agent mode focuses on remote write use cases, let's explain it in more detail.
```